assets/shiny-server.js

```diff
@@ -325,70 +325,149 @@
       }
     };
 
-    // @param reconnect If true, show the reconnecting dialog. If false, show 
-    // disconnected
     this.startReconnect = function(){
-      self._disconnectDialog = true;
-      $('body').addClass('ss-reconnecting');
+      var dialogMsg = '';
+      function setReconnectDialog(msg){
+        // Buffer the msg in case this element isn't visible in the DOM now
+        dialogMsg = msg;
+        $('#ss-connect-dialog').html(msg);
+      }
+
+      // Schedule the display of the disconnect window
+      self._disconnectTimer = setTimeout(function(){
+        debug('Displaying disconnect screen.');
+        $('body').addClass('ss-reconnecting');
+        $('<div id="ss-connect-dialog">'+dialogMsg+'<div class="ss-clearfix"></div></div><div id="ss-gray-out"></div>').appendTo('body');
+      }, 3500);
```

Comment:
I'm a little uneasy about not showing the user any indication that reconnecting happened. Isn't it possible that messages in one direction or the other were lost? I'm aware that we do buffering, but does that cover in-flight messages at the time of the error?

I wonder if the right thing would be to show a reconnecting dialog always and then, on successful reconnect, show "Reconnected to your session" or something for a couple seconds.. which then fades away.